op-node,op-service: Add Fallback Beacon Client

trianglesphere opened the pull request:
**Description**

This splits out a very thin wrapper around the Beacon API into the BeaconClient interface. 
This also adds a simple pool to rotate between different clients on error. This is only used for the blob fetcher. 

This also a beacon archiver to be used as a fallback while still pulling most of the relevant data from the beacon node.

**Tests**

Simple test to validate that the fallback works correctly.

**Metadata**

- Fixes https://github.com/ethereum-optimism/protocol-quest/issues/123

protolambda commented on op-service/sources/l1_beacon_client.go:
```diff
@@ -63,6 +78,93 @@ func (cl *L1BeaconClient) apiReq(ctx context.Context, dest any, reqPath string,
+}
+
+func (cl *FallbackBeaconClient) NodeVersion(ctx context.Context) (string, error) {
+	if r, e := cl.primary.NodeVersion(ctx); e == nil {
+		return r, e
+	}
```
`return r, nil` would be more readable.

Also, can we log the error, so we can tell when the primary node is failing and when we are using the secondary as fallback?

Same for the other API methods below.

sebastianst commented on op-service/sources/l1_beacon_client.go:
```diff
@@ -63,6 +80,106 @@ func (cl *L1BeaconClient) apiReq(ctx context.Context, dest any, reqPath string,
+	l         log.Logger
+}
+
+func NewFallbackClient(a, b client.HTTP, l log.Logger) *FallbackBeaconClient {
```
```suggestion
func NewFallbackBeaconClient(a, b client.HTTP, l log.Logger) *FallbackBeaconClient {
```

sebastianst commented on op-service/sources/l1_beacon_client.go:
```diff
@@ -63,6 +80,106 @@ func (cl *L1BeaconClient) apiReq(ctx context.Context, dest any, reqPath string,
+func (cl *FallbackBeaconClient) BeaconBlobSideCars(ctx context.Context, fetchAllSidecars bool, slot uint64, hashes []eth.IndexedBlobHash) (eth.APIGetBlobSidecarsResponse, error) {
+	if r, e := cl.primary.BeaconBlobSideCars(ctx, fetchAllSidecars, slot, hashes); e == nil {
+		return r, nil
+	} else if strings.Contains(e.Error(), "404") {
```
we could use `http.StatusNotFound` here

sebastianst commented on op-service/sources/l1_beacon_client.go:
```diff
@@ -63,6 +80,106 @@ func (cl *L1BeaconClient) apiReq(ctx context.Context, dest any, reqPath string,
+		return r, nil
+	} else if strings.Contains(e.Error(), "404") {
+		// Log at debug if 404 because this is expected to happen when fetching old blobs.
+		cl.l.Debug("failed to get beacon blob sidecars from primary", "err", e)
```
maybe use a different log message so it's easier to differentiate from the warning in `else`
```suggestion
		// Log at debug if 404 because this is expected to happen when fetching old blobs.
		cl.l.Debug("blob sidecars not found on primary beacon", "err", e)
```

sebastianst commented on op-service/sources/l1_beacon_client.go:
```diff
@@ -63,6 +80,106 @@ func (cl *L1BeaconClient) apiReq(ctx context.Context, dest any, reqPath string,
+	return cl.secondary.BeaconGenesis(ctx)
+}
+
+func (cl *FallbackBeaconClient) BeaconBlobSideCars(ctx context.Context, fetchAllSidecars bool, slot uint64, hashes []eth.IndexedBlobHash) (eth.APIGetBlobSidecarsResponse, error) {
```
This works, but I think it will have worse performance during syncing than an impl that would remember the last working endpoint. For most of the sync it would, every time, query the first, get a 404, then try the second.

Somewhat related, in the `ActiveL2RollupProvider`, we use a slice for the endpoints (well url strings, but similar) and [remember the last working index](https://github.com/ethereum-optimism/optimism/blob/309c9bfd13e2badcdcc26ea16c325ea540fee131/op-service/dial/active_rollup_provider.go#L28).

sebastianst reviewed (commented):
The current blob archiver only supports the `/eth/v1/beacon/blob_sidecars` endpoint, it's not a full backup that provides the other endpoints and doesn't support transparency yet. So we might even embed the primary beacon source in the fallback struct and only do the fallback logic for the `BeaconBlobSideCars` endpoint.

This might also work better when we start remembering the latest beacon endpoint to have blob sidecars, see review comments. If you decide to do that, it might be easier to work with a slice instead of `primary`/`secondary` fields and remember the index of the last working blob endpoint.

sebastianst commented on op-service/sources/l1_beacon_client.go:
```diff
@@ -114,17 +211,10 @@ func (cl *L1BeaconClient) GetBlobSidecars(ctx context.Context, ref eth.L1BlockRe
-
-	var resp eth.APIGetBlobSidecarsResponse
-	if err := cl.apiReq(ctx, &resp, reqPath, reqQuery); err != nil {
+	f := cl.pool.Get()
+	resp, err := f.BeaconBlobSideCars(ctx, cl.cfg.FetchAllSidecars, slot, hashes)
+	if err != nil {
+		cl.pool.MoveToNext()
 		return nil, fmt.Errorf("failed to fetch blob sidecars for slot %v block %v: %w", slot, ref, err)
 	}
```
What do you think about making the fallback mechanism more transparent, ie. trying out the whole pool before returning an error? We could silently fall back and only return an error if all endpoints fail, like the following in its own function
```suggestion
var errs []error
for i := 0; i < cl.pool.Len(); i++ {
	f := cl.pool.Get()
	resp, err := f.BeaconBlobSideCars(ctx, cl.cfg.FetchAllSidecars, slot, hashes)
	if err != nil {
	// log that we're moving to next
		cl.pool.MoveToNext()
		errs = append(errs, fmt.Errorf("failed to fetch blob sidecars for slot %v block %v: %w", slot, ref, err))
	} else {
		return resp, nil
	}
}
return nil, errors.Join(errs)
```

trianglesphere replied:
this works & is likely a tad faster on errors

trianglesphere committed:
op-node,op-service: Add Fallback Beacon Client

This splits out a very thin wrapper around the Beacon API into the
BeaconClient interface. This interface is then implement by two different
structs. The first is a simple wrapper around the HTTP Client. The second
is a FallBack Client which tries two different underlying clients.

This also a beacon archiver to be used as a fallback while still pulling most
of the relevant data from the beacon node.

sebastianst commented on op-node/node/client.go:
```diff
@@ -177,14 +177,20 @@ func (cfg *PreparedL1Endpoint) Check() error {
+func (cfg *L1BeaconEndpointConfig) Setup(ctx context.Context, log log.Logger) (cl sources.BeaconClient, fb []sources.BlobSideCarsFetcher, err error) {
+	a := client.NewBasicHTTPClient(cfg.BeaconAddr, log)
+	if cfg.BeaconArchiverAddr != "" {
+		b := client.NewBasicHTTPClient(cfg.BeaconArchiverAddr, log)
+		fb = append(fb, sources.NewBeaconHTTPClient(b))
```
we could optionally allow specifying multiple archivers if we iterate over `strings.Split(cfg.BeaconArchiverAddr, ",")` here.

trianglesphere replied:
Yea, we could. I'm not sure if it's worth it right now

trianglesphere committed:
Try all clients in the pool

sebastianst reviewed (approved):
LGTM ✨
